[LV] Move condition to VPPartialReductionRecipe::execute

sdesmalen-arm opened the pull request:
This means that VPExpressions will now be constructed for
VPPartialReductionRecipe's when the loop has tail-folding predication.

Note that control-flow (if/else) predication is not yet handled
for partial reductions, because of the way partial reductions
are recognised and built up.

sdesmalen-arm committed:
[LV] NFC: Add new test

This shows that no VPExpression is built for partial reductions
that have some form of predication.

sdesmalen-arm committed:
[LV] Move condition to VPPartialReductionRecipe::execute

This means that VPExpressions will now be constructed for
VPPartialReductionRecipe's when the loop has tail-folding predication.

Note that control-flow (if/else) predication is not yet handled
for partial reductions, because of the way partial reductions
are recognised and built up.

gbossu commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -395,12 +395,18 @@ void VPPartialReductionRecipe::execute(VPTransformState &State) {
+  if (isConditional()) {
+    Value *Cond = State.get(getCondOp());
+    Value *Zero = ConstantInt::get(BinOpVal->getType(), 0);
+    BinOpVal = Builder.CreateSelect(Cond, BinOpVal, Zero);
```
So now we create the `select` directly using an IR builder, as opposed to creating a new recipe for it in `tryToCreatePartialReduction`. Was it that recipe that was preventing tail folding?

SamTebbs33 reviewed (approved):
Looks like a no-brainer to me. I think we should now be able to remove the `m_Select` check in `computeCost`, if you wouldn't mind doing that in this PR as well: https://github.com/llvm/llvm-project/blob/4c3546f964ad0ceba77323b786b94170fd2e3e75/llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp#L319

fhahn reviewed (commented):
Hmm, with this change IIUC we now assume that the cost of partial reduction with and without selects will be the same. Is that correct in all cases?

sdesmalen-arm committed:
Include the cost of the select/predication in the cost of the partial reduction.

In practice this won't make much difference, because VPExpressions already accounts
for the cost of the predication.

sdesmalen-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8171,15 +8171,8 @@ VPRecipeBuilder::tryToCreatePartialReduction(Instruction *Reduction,
+  if (CM.blockNeedsPredicationForAnyReason(Reduction->getParent()))
     Cond = getBlockInMask(Builder.getInsertBlock());
-    VPValue *Zero = Plan.getConstantInt(Reduction->getType(), 0);
-    BinOp = Builder.createSelect(Cond, BinOp, Zero, Reduction->getDebugLoc());
```
yes, my understanding is that is was ignored.

gbossu commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -311,18 +311,27 @@ VPPartialReductionRecipe::computeCost(ElementCount VF,
   // variants are bundled into VPExpressionRecipe.
-  if (!match(Op, m_Select(m_VPValue(), m_VPValue(Op), m_VPValue())) &&
-      !match(Op, m_Mul(m_VPValue(), m_ConstantInt(MulConst)))) {
+  if (!match(Op, m_Mul(m_VPValue(), m_ConstantInt(MulConst)))) {
```
Nit: `MulConst` does not seem to be used. Could that be simplified to `m_ConstantInt(MulConst)`?

sdesmalen-arm replied:
There is only a `m_ConstantInt()` that takes a `uint64_t&`, so it can't be simplified.

gbossu reviewed (approved):
LGTM

fhahn reviewed (commented):
> Hmm, with this change IIUC we now assume that the cost of partial reduction with and without selects will be the same. Is that correct in all cases?

Not sure I missed the response to the comment, but it is still not entirely clear how the cost of the selects is properly accounted for. As far as I can tell, we now create a VPExpressionRecipe with a conditional VPPartialReductionRecipe, but dont account for the reduction being conditional neither when creating VPExpressionRecipe nor in VPExpressionRecipe::computeCost.

As a consequence, aren't we assuming a cost of 1 for VF = 16 for the whole ext/mul/select/reduced in the test case added in `llvm/test/Transforms/LoopVectorize/AArch64/vplan-printing.ll`, while it cannot be lowered to a single partial reduction instruction, with or without SVE? https://llvm.godbolt.org/z/ETas69vna

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -381,12 +390,13 @@ VPPartialReductionRecipe::computeCost(ElementCount VF,
+  return CondCost + Ctx.TTI.getPartialReductionCost(
+                        getOpcode(), InputTypeA, InputTypeB, PhiType, VF,
+                        ExtAType, ExtBType, Opcode, Ctx.CostKind);
+  ;
```
Extraneous ; here
